doc/adr/0012-modules-structure.md

```diff
@@ -0,0 +1,104 @@
+# 12. Modules' structure
+
+Date: 2022-12-28
+
+## Status
+
+Accepted
+
+## Context
+
+The existing structure of tessellation modules was mixed, hence hard to maintain.
+
+- `shared` contains both library extensions and tessellation schema,
+- `keytool` contains key-pair generator and exposes key-related operations,
+- `wallet` contains operations related to creating transactions,
+- `kernel` contains part of state channels (hylomorphism's idea),
+- `sdk` contains all the things shared between Global L0 and DAG L1, but not the $DAG itself,
+- `dag-shared` contains all the things shared between Global L0 and DAG L1 related to $DAG,
+- `core` is the Global L0 module (but also contains state channels that must be extracted to be reused in other modules),
+- `dag-l1` is the DAG L1 module ($DAG token Layer 1).
+
+```mermaid
+graph TD
+    A[Shared]
+    B[Keytool]
+    C[Kernel]
+    D[Wallet]
+    E[SDK]
+    F[DAG Shared]
+    G[Core]
+    H[DAG L1]
+    
+    B --> A
+    C --> A
+    D --> B
+    D --> A
+    E --> C
+    E --> B
+    E --> A
+    E --> F
+    F --> C
+    F --> B
+    F --> A
+    G --> E
+    G --> C
+    G --> B
+    G --> A
+    G --> F
+    H --> F
+    H --> E
+    H --> C
+    H --> B
+    H --> A
+```
+
+Furthermore, dependencies were partially explicitly declared and partially transient. After some time, it became unmaintainable.
+
+The second reason for restructuring modules is the fact that 3rd parties cannot use the SDK module in the current form. It does not expose all required operations, so internal modules were used directly, leaking the project's internal structure and locking further internal changes.
+
+## Decision
+
+To simplify the structure of modules, we will split `shared` into two separate modules - the first containing scala and libraries extensions, the latter containing the tessellation schema. `keytool` module will be splitted into: operations/functions that we will put to `schema` and exposing an app that will be combined into `wallet` module.
+Then, all the things related to node operations would be kept in `node-shared` module (including gossip protocol, consensus, etc.). This leads to emptying the `sdk` module, so we will be able to implement the `sdk` module as a proxy between the external world and the internal tessellation structure. `sdk` module should reexpose all necessary operations to be used by 3rd parties. For example - to create a keypair, global L0 can use the `schema` module directly, but 3rd party will use the `sdk` module that will reexpose keypair creation function. That way, the internal structure of tessellation will be hidden behind the facade.
+
+Modules will be transient - for example: `wallet` does not need to depend on `shared` directly, because it depends on `schema` that depends on `shared`.
```

Comment:
Wait, I think we don't want to have transitive dependencies. If a module wants to use functionality from `shared` then it should depend on `shared`, not just rely on the fact that `schema` depends on `shared`

Reply:
Not sure that we decided about it and that it's set in stone. I wouldn't state that here. We can say that in the new schema graph we don't explicitly mark the transitive dependencies, but it's not yet decided how the dependencies will be handled.